### PR TITLE
[WIP] ci: use gcc-7 and lld

### LINKED ISSUE #4170: Use lld as linker
As a part of tool chain revisit (#4159, #4158) and also as a part of (#3741). We should consider use [LLD](https://lld.llvm.org/) as default linker. LLD is at least twice faster than gold linker, we have more than 300 test target to link so this will improve CI performance.

Options:
1. Use lld everywhere (it works with gcc)
2. Use lld where clang is used (sanitizers)

A sample benchmark of linking `envoy-static` is here:

gold w/o threads (current):
`12.01s user 1.32s system 99% cpu 13.358 total`

gold w/ threads:
`14.24s user 6.97s system 180% cpu 11.762 total`

lld w/o threads: 
`3.63s user 1.15s system 99% cpu 4.805 total`

lld w/ threads:
`8.64s user 11.24s system 709% cpu 2.802 total`

### PR DESCRIPTION
**Pull request by lizan**
Signed-off-by: Lizan Zhou <lizan@tetrate.io>

*Description*:
Another attempt of #5698. Fixes #4170 as well.

*Risk Level*: Low
*Testing*: CI
*Docs Changes*:
*Release Notes*: Added

### CODE REVIEW

File: ci/build_setup.sh [hunk 1]
```diff
@@ -13,6 +13,8 @@ echo "ENVOY_SRCDIR=${ENVOY_SRCDIR}"
 function setup_gcc_toolchain() {
   export CC=gcc
   export CXX=g++
+  [ -z "$CIRCLECI" ] || export BAZEL_BUILD_OPTIONS="${BAZEL_BUILD_OPTIONS} --copt=--param=ggc-min-expand=40"
```
Comment (mattklein123) [inline ci/build_setup.sh]: Can you add a comment on how you figured this param out and what it does?

File: ci/build_setup.sh [hunk 2]
```diff
@@ -62,7 +64,8 @@ export BAZEL="bazel"
 # Not sandboxing, since non-privileged Docker can't do nested namespaces.
 BAZEL_OPTIONS="--package_path %workspace%:${ENVOY_SRCDIR}"
 export BAZEL_QUERY_OPTIONS="${BAZEL_OPTIONS}"
-export BAZEL_BUILD_OPTIONS="--strategy=Genrule=standalone --spawn_strategy=standalone \
+export BAZEL_BUILD_OPTIONS="--action_env=PATH=/bin:/usr/bin:/usr/lib/llvm-7/bin --linkopt=-fuse-ld=lld \
```
Comment (mattklein123) [inline ci/build_setup.sh]: If we are going to link the release build with lld I think we should probably add a release note? I'm not sure that it really matters but it seems like something we should note perhaps? WDYT?

File: test/run_envoy_bazel_coverage.sh [hunk 3]
```diff
@@ -80,7 +80,7 @@ rm "${SRCDIR}"/test/coverage/BUILD
 if [ "$VALIDATE_COVERAGE" == "true" ]
 then
   COVERAGE_VALUE=$(grep -Po 'lines: \K(\d|\.)*' "${COVERAGE_SUMMARY}")
-  COVERAGE_THRESHOLD=98.0
+  COVERAGE_THRESHOLD=97.5
```
Comment (mattklein123) [inline test/run_envoy_bazel_coverage.sh]: This is unfortunate. I guess there is no way around this in terms of some other gcovr exclusion type?

File: ci/build_setup.sh (hunk 1 above)
Comment (lizan) [reply ci/build_setup.sh]: added

File: ci/build_setup.sh (hunk 2 above)
Comment (lizan) [reply ci/build_setup.sh]: yeah sure, added to version_history along with compiler update.

File: test/run_envoy_bazel_coverage.sh (hunk 3 above)
Comment (lizan) [reply test/run_envoy_bazel_coverage.sh]: Yeah with new GCC it includes more line, which led this. I think we may bring back this to 98.0 without much hack, but it is too much for this PR. I didn't try much with hacking gcovr, @htuch any idea?

**Comment by lizan** [comment]
hmm, I think the googletest update let compiler uses more memory and in conjunction with gcc-7 it doesn't work anymore... :/

**Comment by mattklein123** [comment]
@lizan I had one small thought: You could bazel build `//source` with all CPUs, and then use less jobs for building tests, and then maybe use all CPUs for running tests. This might workaround the memory issues without using less jobs for everything?

File: test/run_envoy_bazel_coverage.sh (hunk 3 above)
Comment (htuch) [reply test/run_envoy_bazel_coverage.sh]: With bazel native coverage, I think it will drop anyway. We're still trying to debug this thought CC @iirina.

**Comment by lizan** [comment]
In favor of #5698, will make lld change later.

**Pull request closed without merging**